scipy/optimize/_minpack_py.py

```diff
@@ -800,8 +815,12 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
         raise ValueError("Method 'lm' only works for unconstrained problems. "
                          "Use 'trf' or 'dogbox' instead.")
 
-    # optimization may produce garbage for float32 inputs, cast them to float64
+    if nan_policy is None:  # If nan_policy is not specified explicitly
+        check_finite = True  # check finite for backward compatibility
+    elif check_finite is None:  # if only nan_policy is specified explicitly
+        check_finite = False  # not check finite
```

Comment:
This is backward incompatible because it overrides whatever the user passes for `check_finite` when `nan_policy` is unspecified.  I see that my comment (https://github.com/scipy/scipy/pull/16961#discussion_r1041530008) was unclear, but this is what I had in mind:

```suggestion
    if check_finite is None:
        check_finite = False if nan_policy is not None else True
```